### ISSUE TITLE
Context lifetime bound to garbage collection

### ISSUE THREAD

**Issue by silbe** [nipo/ausb#4]
`Context` has no public API for shutting it down and detaching `ContextNotifier` from the asyncio loop. This makes working with ausb in projects where the asyncio loop gets replaced repeatedly (e.g. when using [pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio)) quite cumbersome. All references to `Context` need to be controlled carefully, using weak references in some places where "hidden" references cannot be avoided (e.g. pytest fixture caching). It also takes quite a lot of additional steps to ensure `ContextNotifier._work()` is actually finished (and thus detached from the asyncio loop) before the asyncio loop may be shut down.

On top of that, ausb triggers warnings in usb1:

```
/usr/lib/python3/dist-packages/usb1/__init__.py:2071: DeprecationWarning: Use "with USBContext() as context:" for safer cleanup on interpreter shutdown. See also USBContext.open().
```

I have some working PoC code for turning `ContextNotifier.close()` into a coroutine (so that it can wait for `ContextNotifier._work()` to return), adding a coroutine `Context.close()` and having the `Context` finalizer use `loop.run_until_complete()` (so that existing programs continue to work without explicit `close()`). This fixes at least the immediate issue of not being able to explicitly shut down `Context` and waiting for it to be detached from the asyncio loop. `__aenter__()` / `__aexit__()` would be natural additions on top of that to make it work similar to the recommended way of using `usb1.USBContext`.

The usb1 warning can probably be avoided by invoking `usb1.USBContext.__enter__()` in `Context.__init__()` and `usb1.USBContext.__exit__()` in `Context.close()`. Haven't tried that out yet.

Before cleaning up my PoC and sending a PR I'd like to make sure we agree on the approach.

**Comment by nipo** [comment]
I agree context manager would be better for most simple usages. Nonetheless, I more often use ausb in a plugin fashion where device enumeration/registration/usage is done as part of a set of interchangeable  backends (not all USB-related). Used this way, usb-specific code is deep-down in the call tree and returns early, making context manager useless. (I'm still looking for clean ways to implement such pattern).

So I clearly see added value for implementing context manager, but this should just be syntactic sugar over a generic method with classical context instances. I agree shifting close() to a coroutine is mandatory, as it's the only clean way to block caller until "background thread" got killed.

Overall, we agree.
